src/win/fs.c

```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
 
 
 static void fs__rename(uv_fs_t* req) {
-  if (!MoveFileExW(req->file.pathw, req->fs.info.new_pathw, MOVEFILE_REPLACE_EXISTING)) {
+  int tries;
+  int sys_errno;
+  int result;
+  int try_rmdir;
+  WCHAR *src, *dst;
+  DWORD src_attrib, dst_attrib;
+
+  src = req->file.pathw;
+  dst = req->fs.info.new_pathw;
+  try_rmdir = 0;
+
+  /* Do some checks to fail early. */
+  src_attrib = GetFileAttributesW(src);
+  if (src_attrib == INVALID_FILE_ATTRIBUTES) {
     SET_REQ_WIN32_ERROR(req, GetLastError());
     return;
   }
+  dst_attrib = GetFileAttributesW(dst);
+  if (dst_attrib != INVALID_FILE_ATTRIBUTES) {
+    if (dst_attrib & FILE_ATTRIBUTE_READONLY) {
+      req->result = UV_EPERM;
+      return;
+    }
+    /* Renaming folder to a folder name that already exist will fail on
+     * Windows. We will try to delete target folder first.
+     */
+    if (src_attrib & FILE_ATTRIBUTE_DIRECTORY &&
+        dst_attrib & FILE_ATTRIBUTE_DIRECTORY) {
+        try_rmdir = 1;
+    }
+  }
 
-  SET_REQ_RESULT(req, 0);
+  /* Sometimes an antivirus or indexing software can lock the target or the
+   * source file/directory. This is annoying for users, in such cases we will
+   * retry couple of times with some delay before failing.
+   */
+  for (tries = 0; tries < UV__RENAME_RETRIES; ++tries) {
+    if (tries > 0) {
+      Sleep(UV__RENAME_WAIT);
+    }
+    if (try_rmdir) {
+      if (_wrmdir(dst) == 0) {
+        result = 0;
+      } else {
+        result = uv_translate_sys_error(_doserrno);
+      }
+      switch (result)
+      {
+      case 0:
+      case UV_ENOENT:
+        /* Folder removed or did not exist at all. */
+        try_rmdir = 0;
```

Comment:
When copying directories, mv might have copied a few files and subdirectories before failing. In that case, shouldn't we try to delete the destination again?

Reply:
The directory is renamed, it happens at once.